File: src/python/inputs-complex/zlib-1.3.1/individual-funcs/compress2.c
int ZEXPORT compress2(Bytef *dest, uLongf *destLen, const Bytef *source,
                      uLong sourceLen, int level) {
    z_stream stream;
    int err;
    const uInt max = (uInt)-1;
    uLong left;

    left = *destLen;
    *destLen = 0;

    stream.zalloc = (alloc_func)0;
    stream.zfree = (free_func)0;
    stream.opaque = (voidpf)0;

    err = deflateInit(&stream, level);
    if (err != Z_OK) return err;

    stream.next_out = dest;
    stream.avail_out = 0;
    stream.next_in = (z_const Bytef *)source;
    stream.avail_in = 0;

    do {
        if (stream.avail_out == 0) {
            stream.avail_out = left > (uLong)max ? max : (uInt)left;
            left -= stream.avail_out;
        }
        if (stream.avail_in == 0) {
            stream.avail_in = sourceLen > (uLong)max ? max : (uInt)sourceLen;
            sourceLen -= stream.avail_in;
        }
        err = deflate(&stream, sourceLen ? Z_NO_FLUSH : Z_FINISH);
    } while (err == Z_OK);

    *destLen = stream.total_out;
    deflateEnd(&stream);
    return err == Z_STREAM_END ? Z_OK : err;
}

File: src/python/inputs-complex/zlib-1.3.1/individual-funcs/compress2.rs
use zlib::{Deflate, DeflateOptions};

fn compress2(dest: &mut [u8], dest_len: &mut usize, source: &[u8], source_len: usize, level: i32) -> i32 {
    let mut stream = Deflate::new(level).unwrap();
    let mut left = *dest_len;
    *dest_len = 0;

    loop {
        if left == 0 {
            left = source_len.min(usize::max_value() as usize);
        }

        let result = stream.deflate(source, left, Z_NO_FLUSH);
        if result != Z_OK {
            break result;
        }

        *dest_len += result.total_out();
        source_len -= result.avail_in() as usize;
    }
}